### DISCUSSION TITLE
How can I properly encode and hash a message that contains a struct array in accordance to EIP712 using Viem.

### DISCUSSION THREAD

**Discussion by 0xFloop** [wevm/viem#1077] [Question] [answered]
I cannot seem to figure out how to properly hash a struct (MainStruct in the example) that contains a struct array within it (SubStruct[] in the example) so that it produces the same hash as that generated by Viem's hashTypedData function. I have written out three different sol hashing options in the README that I have tried. [Here](https://github.com/0xFloop/eip712-struct-array/) is a reproduction repo. Thank you very very much for any help and let me know any more info you may need!

**Comment by 0xFloop** [comment]
@tmm  any chance you can lend a hand with this? would greatly appreciate your help.

**Comment by 0xFloop** [comment, marked as answer]
Full solution code [here](https://github.com/0xFloop/eip712-struct-array).

TLDR: 
eip712 states in regards to hashing arrays/structs that "The array values are encoded as the keccak256 hash of the concatenated encodeData of their contents (i.e. the encoding of SomeType[5] is identical to that of a struct containing five members of type SomeType). The struct values are encoded recursively as hashStruct(value)"

My error came in assuming that these values would be concatenated using the abi.encode method. However, you are meant to use the abi.encodePacked method. This definitely confused me as to why data should be packed here but not in other cases, like within hashStruct for example. Anywho, thats the resolution.

**Reply by dcapitator** [reply]
> Full solution code [here](https://github.com/0xFloop/eip712-struct-array).
> 
> TLDR: eip712 states in regards to hashing arrays/structs that "The array values are encoded as the keccak256 hash of the concatenated encodeData of their contents (i.e. the encoding of SomeType[5] is identical to that of a struct containing five members of type SomeType). The struct values are encoded recursively as hashStruct(value)"
> 
> My error came in assuming that these values would be concatenated using the abi.encode method. However, you are meant to use the abi.encodePacked method. This definitely confused me as to why data should be packed here but not in other cases, like within hashStruct for example. Anywho, thats the resolution.

i guess i have similar issue but ur solution didnt solve mine.

onchain orderHash looks like this :
```
            keccak256(
                abi.encode(
                    ORDER_TYPEHASH,
                    order.registry,
                    order.maker,
                    order.executer,
                    order.staticTarget,
                    order.staticSelector,
                    keccak256(order.staticExtradata),
                    order.maximumFill,
                    order.extraData
                )
            );
```

and there is difference between ethers v6 and viem with hashing structs 

Viem
```
  const orderHash = hashTypedData({
    types,
    primaryType: 'Order',
    message,
  })
```

Ethers:
```
  const orderHash = TypedDataEncoder.hashStruct(
    'Order',
    types,
    message
  )
```

ethers gives same hash returned from contract . but viem returns a different hash.. 
types , message typehash are all correct. Adding domain to viem hash didnt work and its not necessary.

Any idea why those 2 functions return different hashes ??
